docs/onboarding/09-exposing-twins-over-graphql.md

```diff
@@ -0,0 +1,46 @@
+# Chapter 9: Exposing Digital Twins over GraphQL
+
+- [Why GraphQL?](https://www.apollographql.com/docs/intro/benefits/) - Explains the value proposition of GraphQL
+- [GraphQL Tutorial](https://www.apollographql.com/docs/tutorial/introduction/) - Walks through coding a GraphQL service. **STOP after [Chapter 4 (Write Mutation Resolvers)](https://www.apollographql.com/docs/tutorial/mutation-resolvers/)**. Apollo Explorer is not necessary to complete the tutorial. Browsing to localhost address provides means to execute requests.
+
+## Key Points
+
+- GraphQL is a declarative model for exposing data from multiple sources over a common protocol.
+- Data Sources facilitate operations against a specific source of data such as a database or api.
+- Resolvers leverage one or more data sources to fulfill a requested operation
+- Schema defines both the model/state (e.g. Room, Floor) exposed as well as the supported operations (GetRoomsByFloor)
+- Clients must leverage an SDK that translates GraphQL operation into a protocol specific implementation (i.e. HTTP) as opposed to directly invoking an HTTP endpoint.
+- GraphQL is a standard specification while Apollo Server is one implementation of the GraphQL standard.
+
+> Since Azure Digital Twins also follows a declarative model for defining twins and relationships and, twins are likely to co-exist with other non-twin application data; GraphQL is a natural fit for exposing the aggregate of this information over a common protocol.
+
+## Learning Exercises
+
+### Prerequisites
+
+Javascript/Node will need to be installed/used to complete the exercises using Apollo GraphQL framework. VS Code is suggested IDE for completing these challenges.
+
+1. [NodeJS 14.16.1](https://nodejs.org/dist/v14.16.1/node-v14.16.1-x64.msi)
+2. [Azure Digital Twins Data Plane Javascript SDK](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/digitaltwins/digital-twins-core)
+3. [Apollo Server](https://www.npmjs.com/package/apollo-server)
+
+### Exercises
+
+1. Create a GraphQL service that exposes twins of at least one model. For example, given a `Floor` model, implement a `GetFloors` query that returns all `Floor` twins.

```

Comment:
That has been a big jump for me (I am actually still working on the exercise). Maybe we can include an intro in this section (or another step), to provide more guidance on how to set it up. I also found [these examples](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/digitaltwins/digital-twins-core/samples/v1/javascript) very helpful (the link was provided above but I didn't find them immediately).

Reply:
Completely agreed! Went through the GraphQL tutorial, was pretty straightforward -- and then got to the first exercise and wasn't sure where to start. Requires too much research on the upskiller's side -- @norelina's suggestions are great here.

Reply:
I decomposed the exercises into progression towards getting a running server as opposed to that being step 1. Now it will follow this progression

1. Setup the ADT SDK for Javascript and run a sample query to get twins
2. Setup GQL schema that mirrors the ADT model
3. Implement a GQL Data Source
4. Implement a GQL Resolver
5. Tie it all together into a GQL Server and test running GQL Queries

I would like these exercises to serve as a gauge as to how much of a learning curve GQL + ADT is going to be for the team. If everyone is stuck day 1 then that can surface that the team has a much steeper learning curve than we may have anticipated. For that reason, i want to strike a balance between guidance and struggle. The content modifications I believe decompose a step by step approach to solve the problem. However, I do not want to just provide pre-coded data sources, resolvers, etc as I want to see if the team can put these together on their own and connect it to ADT.